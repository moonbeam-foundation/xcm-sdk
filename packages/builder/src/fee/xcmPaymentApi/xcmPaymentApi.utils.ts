import type { AnyChain } from '@moonbeam-network/xcm-types';
import type { FeeConfigBuilderParams } from '../FeeBuilder.interfaces';
import {
  getBuyExecutionInstruction,
  getClearOriginInstruction,
  getDepositAssetInstruction,
  getDescendOriginInstruction,
  getReserveAssetDepositedInstruction,
  getSetTopicInstruction,
  getUniversalOriginInstruction,
  getWithdrawAssetInstruction,
} from '../FeeBuilder.utils';
import type { CreateXcmFeeBuilderProps } from './xcmPaymentApi';

export interface GetVersionedAssetsProps extends CreateXcmFeeBuilderProps {
  params: FeeConfigBuilderParams;
}

export async function getVersionedAssets({
  getVersionedFeeAsset,
  getVersionedTransferAsset,
  options,
  params,
}: GetVersionedAssetsProps): Promise<[object[], object]> {
  const { asset: transferAsset, feeAsset } = params;
  const versionedFeeAssetId = await getVersionedFeeAsset(params);

  const assets = [versionedFeeAssetId];

  if (feeAsset !== transferAsset && getVersionedTransferAsset) {
    const versionedTransferAssetId = await getVersionedTransferAsset(params);

    if (options.shouldTransferAssetPrecedeFeeAsset) {
      assets.unshift(versionedTransferAssetId);
    } else {
      assets.push(versionedTransferAssetId);
    }
  }

  return [assets, versionedFeeAssetId];
}

interface GetInstructionsProps {
  isAssetReserveChain: boolean;
  assets: object[];
  versionedFeeAssetId: object;
  address: string;
  source: AnyChain;
  isEcosystemBridge?: boolean;
}

export function getInstructions({
  isAssetReserveChain,
  assets,
  versionedFeeAssetId,
  address,
  source,
  isEcosystemBridge,
}: GetInstructionsProps) {
  const instructions = [
    isAssetReserveChain
      ? getWithdrawAssetInstruction(assets)
      : getReserveAssetDepositedInstruction(assets),
    getClearOriginInstruction(),
    getBuyExecutionInstruction(versionedFeeAssetId),
    getDepositAssetInstruction(address, assets),
    getSetTopicInstruction(),
  ];

  if (isEcosystemBridge) {
    return [
      getUniversalOriginInstruction(source),
      getDescendOriginInstruction(source),
      ...instructions,
    ];
  }

  return instructions;
}
