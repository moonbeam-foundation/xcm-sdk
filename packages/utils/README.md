The Utils package contains the utility functions for the Moonbeam XCM SDK and MRL SDK.

# Documentation for the Moonbeam XCM SDK

## v3 (current)

### Usage

- [XCM SDK](https://moonbeam-foundation.github.io/xcm-sdk/latest/example-usage/xcm)
- [MRL SDK](https://moonbeam-foundation.github.io/xcm-sdk/latest/example-usage/mrl)

### References

- [XCM SDK](https://moonbeam-foundation.github.io/xcm-sdk/latest/reference/xcm)
- [MRL SDK](https://moonbeam-foundation.github.io/xcm-sdk/latest/reference/mrl)

## v2 (previous)

- [usage](https://moonbeam-foundation.github.io/xcm-sdk/v2/example-usage)
- [references](https://moonbeam-foundation.github.io/xcm-sdk/v2/reference/interfaces)