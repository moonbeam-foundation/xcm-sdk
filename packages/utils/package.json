{
  "name": "@moonbeam-network/xcm-utils",
  "version": "2.2.5",
  "description": "Moonbeam XCM utilities",
  "scripts": {
    "build": "tsup",
    "dev": "tsup --watch",
    "link": "pnpm ln --global",
    "typecheck": "tsc --noEmit",
    "test": "vitest --run",
    "test:watch": "vitest",
    "test:update": "vitest -u"
  },
  "author": "moonbeam-foundation",
  "license": "MIT",
  "repository": {
    "directory": "packages/utils",
    "type": "git",
    "url": "git+https://github.com/moonbeam-foundation/xcm-sdk.git"
  },
  "keywords": ["moonbeam", "moonriver", "xcm", "utils"],
  "bugs": {
    "url": "https://github.com/moonbeam-foundation/xcm-sdk/issues"
  },
  "homepage": "https://moonbeam-foundation.github.io/xcm-sdk/latest",
  "files": ["build"],
  "type": "module",
  "exports": {
    "import": "./build/index.mjs",
    "types": "./build/index.d.ts",
    "default": "./build/index.mjs"
  },
  "types": "./build/index.d.ts",
  "main": "./build/index.mjs",
  "dependencies": {
    "big.js": "^6.2.1",
    "lru-cache": "^10.2.2"
  },
  "devDependencies": {
    "@types/big.js": "^6.2.2"
  },
  "peerDependencies": {
    "@polkadot/api": "14.3.1",
    "@polkadot/apps-config": "0.146.1",
    "@polkadot/util": "13.2.3",
    "@polkadot/util-crypto": "13.2.3"
  }
}
