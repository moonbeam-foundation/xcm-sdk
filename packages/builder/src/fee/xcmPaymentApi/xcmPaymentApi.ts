import { SubstrateCallConfig } from '../../types/substrate/SubstrateCallConfig';
import type {
  FeeConfigBuilder,
  FeeConfigBuilderParams,
  GetVersionedAssetId,
  XcmPaymentFeeProps,
} from '../FeeBuilder.interfaces';
import {
  getBuyExecutionInstruction,
  getClearOriginInstruction,
  getDepositAssetInstruction,
  getFeeForXcmInstructionsAndAsset,
  getReserveAssetDepositedInstruction,
  getSetTopicInstruction,
  getVersionedAssetId,
  getWithdrawAssetInstruction,
} from '../FeeBuilder.utils';
import {
  BuildVersionedAsset,
  QueryVersionedAsset,
} from '../VersionedAssetBuilder';
import { getInstructions } from './xcmPaymentApi.utils';
import { getVersionedAssets } from './xcmPaymentApi.utils';

export function xcmPaymentApi() {
  const localMethods = {
    fromHere: (options: XcmPaymentFeeProps) =>
      createXcmFeeBuilder({
        getVersionedFeeAsset: () =>
          BuildVersionedAsset().fromHere(options.parents),
        options,
      }),

    fromHereAndGeneralIndex: (options: XcmPaymentFeeProps) =>
      createXcmFeeBuilder({
        getVersionedFeeAsset: () => BuildVersionedAsset().fromHere(),
        getVersionedTransferAsset: ({ asset }) =>
          BuildVersionedAsset().fromPalletInstanceAndGeneralIndex(asset),
        options,
      }),

    fromPalletInstanceAndGeneralIndex: (options: XcmPaymentFeeProps) =>
      createXcmFeeBuilder({
        getVersionedFeeAsset: ({ feeAsset }) =>
          BuildVersionedAsset().fromPalletInstanceAndGeneralIndex(feeAsset),
        options,
      }),

    fromHereAndSourceGeneralIndex: (options: XcmPaymentFeeProps) =>
      createXcmFeeBuilder({
        getVersionedFeeAsset: () => BuildVersionedAsset().fromHere(),
        getVersionedTransferAsset: ({ source, asset }) =>
          BuildVersionedAsset()
            .fromSource()
            .palletInstanceAndGeneralIndex(source, asset),
        options,
      }),

    fromGeneralIndex: (options: XcmPaymentFeeProps) =>
      createXcmFeeBuilder({
        getVersionedFeeAsset: ({ feeAsset }) =>
          BuildVersionedAsset().fromGeneralIndex(feeAsset),
        options,
      }),

    fromPalletInstance: (options: XcmPaymentFeeProps) =>
      createXcmFeeBuilder({
        getVersionedFeeAsset: ({ feeAsset }) =>
          BuildVersionedAsset().fromPalletInstance(feeAsset),
        options,
      }),

    fromPalletInstanceAndAccountKey20: (options: XcmPaymentFeeProps) =>
      createXcmFeeBuilder({
        getVersionedFeeAsset: ({ feeAsset }) =>
          BuildVersionedAsset().fromPalletInstance(feeAsset),
        getVersionedTransferAsset: ({ asset }) =>
          BuildVersionedAsset().fromAccountKey20(asset),
        options,
      }),

    fromGlobalConsensus: (options: XcmPaymentFeeProps) =>
      createXcmFeeBuilder({
        getVersionedFeeAsset: ({ feeAsset }) =>
          BuildVersionedAsset().fromGlobalConsensus(feeAsset),
        options,
      }),
  };

  const queryMethods = {
    fromCurrencyIdToLocations: (options: XcmPaymentFeeProps) =>
      createXcmFeeBuilder({
        getVersionedFeeAsset: async ({ source, feeAsset, api }) => {
          return await QueryVersionedAsset().fromCurrencyIdToLocations(
            feeAsset,
            api,
          );
        },
        options,
      }),
  };

  const sourceMethods = {
    fromSourceAccountKey20: (options: XcmPaymentFeeProps) =>
      createXcmFeeBuilder({
        getVersionedFeeAsset: ({ source, asset }) =>
          BuildVersionedAsset().fromSource().accountKey20(source, asset),
        options,
      }),

    fromSourceGeneralIndex: (options: XcmPaymentFeeProps) =>
      createXcmFeeBuilder({
        getVersionedFeeAsset: ({ source, feeAsset }) =>
          BuildVersionedAsset().fromSource().generalIndex(source, feeAsset),
        options,
      }),

    fromSourcePalletInstance: (options: XcmPaymentFeeProps) =>
      createXcmFeeBuilder({
        getVersionedFeeAsset: ({ source, feeAsset }) =>
          BuildVersionedAsset().fromSource().palletInstance(source, feeAsset),
        options,
      }),

    fromSourceGlobalConsensus: (options: XcmPaymentFeeProps) =>
      createXcmFeeBuilder({
        getVersionedFeeAsset: ({ source, feeAsset }) => {
          const sourceAsset = source.getChainAsset(feeAsset);
          return BuildVersionedAsset().fromGlobalConsensus(sourceAsset);
        },
        options,
      }),
  };

  const legacyMethods = {
    // TODO deprecate this after applying to asset migration
    xcmPaymentFee: ({
      isAssetReserveChain,
      shouldTransferAssetPrecedeFeeAsset = false,
    }: XcmPaymentFeeProps): FeeConfigBuilder => ({
      build: ({
        address,
        api,
        asset,
        destination,
        feeAsset,
      }: FeeConfigBuilderParams) =>
        new SubstrateCallConfig({
          api,
          call: async (): Promise<bigint> => {
            const versionedFeeAssetId = await getVersionedAssetId(
              api,
              feeAsset,
              destination,
            );
            const versionedTransferAssetId = await getVersionedAssetId(
              api,
              asset,
              destination,
            );

            const versionedAssets = shouldTransferAssetPrecedeFeeAsset
              ? [versionedTransferAssetId, versionedFeeAssetId]
              : [versionedFeeAssetId, versionedTransferAssetId];

            const assets =
              feeAsset === asset ? [versionedFeeAssetId] : versionedAssets;

            const instructions = [
              isAssetReserveChain
                ? getWithdrawAssetInstruction(assets)
                : getReserveAssetDepositedInstruction(assets),
              getClearOriginInstruction(),
              getBuyExecutionInstruction(versionedFeeAssetId),
              getDepositAssetInstruction(address, assets),
              getSetTopicInstruction(),
            ];

            return getFeeForXcmInstructionsAndAsset(
              api,
              instructions,
              versionedFeeAssetId,
            );
          },
        }),
    }),
  };

  return {
    ...localMethods,
    ...sourceMethods,
    ...queryMethods,
    ...legacyMethods,
  };
}

export interface CreateXcmFeeBuilderProps {
  getVersionedFeeAsset: GetVersionedAssetId;
  getVersionedTransferAsset?: GetVersionedAssetId;
  options: XcmPaymentFeeProps;
}

const createXcmFeeBuilder = ({
  getVersionedFeeAsset,
  getVersionedTransferAsset,
  options,
}: CreateXcmFeeBuilderProps): FeeConfigBuilder => ({
  build: (params: FeeConfigBuilderParams) =>
    new SubstrateCallConfig({
      api: params.api,
      call: async (): Promise<bigint> => {
        const [assets, versionedFeeAssetId] = await getVersionedAssets({
          getVersionedFeeAsset,
          getVersionedTransferAsset,
          options,
          params,
        });

        const instructions = getInstructions({
          isAssetReserveChain: options.isAssetReserveChain,
          assets,
          versionedFeeAssetId,
          address: params.address,
        });

        return getFeeForXcmInstructionsAndAsset(
          params.api,
          instructions,
          versionedFeeAssetId,
        );
      },
    }),
});
