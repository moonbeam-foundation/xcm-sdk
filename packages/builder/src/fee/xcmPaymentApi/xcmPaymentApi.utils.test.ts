import type { ApiPromise } from '@polkadot/api';
import { describe, expect, it } from 'vitest';
import {
  feeBuilderParamsMock,
  feeBuilderParamsMockDifferentAsset,
} from '../../../fixtures';
import { getInstructions, getVersionedAssets } from './xcmPaymentApi.utils';
import type { GetVersionedAssetsProps } from './xcmPaymentApi.utils';
const mockApi = {
  isReady: Promise.resolve(true),
  isReadyOrError: Promise.resolve(true),
  clone: () => mockApi,
} as unknown as ApiPromise;

describe('xcmPaymentApi.utils', () => {
  describe('getVersionedAssets', () => {
    const mockVersionedFeeAsset = { parents: 1, interior: 'Here' };
    const mockVersionedTransferAsset = {
      parents: 1,
      interior: {
        X1: [{ PalletInstance: 3 }],
      },
    };

    it('should return only fee asset when assets are the same', async () => {
      const props: GetVersionedAssetsProps = {
        getVersionedFeeAsset: async () => mockVersionedFeeAsset,
        getVersionedTransferAsset: async () => mockVersionedTransferAsset,
        options: {
          isAssetReserveChain: true,
          shouldTransferAssetPrecedeFeeAsset: false,
        },
        params: {
          ...feeBuilderParamsMock,
        },
      };

      const [assets, versionedFeeAssetId] = await getVersionedAssets(props);

      expect(assets).toHaveLength(1);
      expect(assets).toContain(mockVersionedFeeAsset);
      expect(versionedFeeAssetId).toBe(mockVersionedFeeAsset);
      expect(assets).toMatchSnapshot();
    });

    it('should put transfer asset first when shouldTransferAssetPrecedeFeeAsset is true', async () => {
      const props = {
        getVersionedFeeAsset: async () => mockVersionedFeeAsset,
        getVersionedTransferAsset: async () => mockVersionedTransferAsset,
        options: {
          isAssetReserveChain: true,
          shouldTransferAssetPrecedeFeeAsset: true,
        },
        params: {
          ...feeBuilderParamsMockDifferentAsset,
        },
      };

      const [assets, versionedFeeAssetId] = await getVersionedAssets(props);

      expect(assets).toHaveLength(2);
      expect(assets[0]).toBe(mockVersionedTransferAsset);
      expect(assets[1]).toBe(mockVersionedFeeAsset);
      expect(versionedFeeAssetId).toBe(mockVersionedFeeAsset);
      expect(assets).toMatchSnapshot();
    });

    it('should put fee asset first when shouldTransferAssetPrecedeFeeAsset is false', async () => {
      const props = {
        getVersionedFeeAsset: async () => mockVersionedFeeAsset,
        getVersionedTransferAsset: async () => mockVersionedTransferAsset,
        options: {
          isAssetReserveChain: true,
          shouldTransferAssetPrecedeFeeAsset: false,
        },
        params: {
          ...feeBuilderParamsMockDifferentAsset,
        },
      };

      const [assets, versionedFeeAssetId] = await getVersionedAssets(props);

      expect(assets).toHaveLength(2);
      expect(assets[0]).toBe(mockVersionedFeeAsset);
      expect(assets[1]).toBe(mockVersionedTransferAsset);
      expect(versionedFeeAssetId).toBe(mockVersionedFeeAsset);
      expect(assets).toMatchSnapshot();
    });
  });

  describe('getInstructions', () => {
    const mockAddress = '0x98891e5FD24Ef33A488A47101F65D212Ff6E650E';
    const mockAssets = [{ id: 1 }, { id: 2 }];
    const mockFeeAssetId = { id: 1 };

    it('should return withdraw asset instruction when isAssetReserveChain is true', () => {
      const instructions = getInstructions({
        isAssetReserveChain: true,
        assets: mockAssets,
        versionedFeeAssetId: mockFeeAssetId,
        address: mockAddress,
      });

      expect(instructions).toHaveLength(5);
      expect(instructions).toMatchSnapshot();
    });

    it('should return reserve asset deposited instruction when isAssetReserveChain is false', () => {
      const instructions = getInstructions({
        isAssetReserveChain: false,
        assets: mockAssets,
        versionedFeeAssetId: mockFeeAssetId,
        address: mockAddress,
      });

      expect(instructions).toHaveLength(5);
      expect(instructions).toMatchSnapshot();
    });

    it('should maintain correct instruction order', () => {
      const instructions = getInstructions({
        isAssetReserveChain: true,
        assets: mockAssets,
        versionedFeeAssetId: mockFeeAssetId,
        address: mockAddress,
      });

      expect(instructions).toMatchSnapshot();
    });
  });
});
