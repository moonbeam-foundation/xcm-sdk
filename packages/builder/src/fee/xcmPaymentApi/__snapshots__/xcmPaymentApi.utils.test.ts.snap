// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`xcmPaymentApi.utils > getInstructions > should handle isEcosystemBridge true with different asset reserve chain settings 1`] = `
[
  {
    "UniversalOrigin": {
      "GlobalConsensus": {
        "ByGenesis": undefined,
      },
    },
  },
  {
    "DescendOrigin": {
      "X1": [
        {
          "Parachain": 1001,
        },
      ],
    },
  },
  {
    "WithdrawAsset": [
      {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 1,
        },
      },
      {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 2,
        },
      },
    ],
  },
  {
    "ClearOrigin": "NULL",
  },
  {
    "BuyExecution": {
      "fees": {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 1,
        },
      },
      "weight_limit": {
        "Unlimited": "NULL",
      },
    },
  },
  {
    "DepositAsset": {
      "assets": {
        "Wild": {
          "AllCounted": 2,
        },
      },
      "beneficiary": {
        "interior": {
          "X1": [
            {
              "AccountKey20": {
                "key": "0x98891e5FD24Ef33A488A47101F65D212Ff6E650E",
                "network": null,
              },
            },
          ],
        },
        "parents": 0,
      },
      "max_assets": 0,
    },
  },
  {
    "SetTopic": "0xabcdef1234567890fedcba0987654321abcdef1234567890fedcba0987654321",
  },
]
`;

exports[`xcmPaymentApi.utils > getInstructions > should handle isEcosystemBridge true with different asset reserve chain settings 2`] = `
[
  {
    "UniversalOrigin": {
      "GlobalConsensus": {
        "ByGenesis": undefined,
      },
    },
  },
  {
    "DescendOrigin": {
      "X1": [
        {
          "Parachain": 1001,
        },
      ],
    },
  },
  {
    "ReserveAssetDeposited": [
      {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 1,
        },
      },
      {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 2,
        },
      },
    ],
  },
  {
    "ClearOrigin": "NULL",
  },
  {
    "BuyExecution": {
      "fees": {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 1,
        },
      },
      "weight_limit": {
        "Unlimited": "NULL",
      },
    },
  },
  {
    "DepositAsset": {
      "assets": {
        "Wild": {
          "AllCounted": 2,
        },
      },
      "beneficiary": {
        "interior": {
          "X1": [
            {
              "AccountKey20": {
                "key": "0x98891e5FD24Ef33A488A47101F65D212Ff6E650E",
                "network": null,
              },
            },
          ],
        },
        "parents": 0,
      },
      "max_assets": 0,
    },
  },
  {
    "SetTopic": "0xabcdef1234567890fedcba0987654321abcdef1234567890fedcba0987654321",
  },
]
`;

exports[`xcmPaymentApi.utils > getInstructions > should maintain correct instruction order 1`] = `
[
  {
    "WithdrawAsset": [
      {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 1,
        },
      },
      {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 2,
        },
      },
    ],
  },
  {
    "ClearOrigin": "NULL",
  },
  {
    "BuyExecution": {
      "fees": {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 1,
        },
      },
      "weight_limit": {
        "Unlimited": "NULL",
      },
    },
  },
  {
    "DepositAsset": {
      "assets": {
        "Wild": {
          "AllCounted": 2,
        },
      },
      "beneficiary": {
        "interior": {
          "X1": [
            {
              "AccountKey20": {
                "key": "0x98891e5FD24Ef33A488A47101F65D212Ff6E650E",
                "network": null,
              },
            },
          ],
        },
        "parents": 0,
      },
      "max_assets": 0,
    },
  },
  {
    "SetTopic": "0xabcdef1234567890fedcba0987654321abcdef1234567890fedcba0987654321",
  },
]
`;

exports[`xcmPaymentApi.utils > getInstructions > should prepend universal origin and descend origin instructions when isEcosystemBridge is true 1`] = `
[
  {
    "UniversalOrigin": {
      "GlobalConsensus": {
        "ByGenesis": undefined,
      },
    },
  },
  {
    "DescendOrigin": {
      "X1": [
        {
          "Parachain": 1001,
        },
      ],
    },
  },
  {
    "ReserveAssetDeposited": [
      {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 1,
        },
      },
      {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 2,
        },
      },
    ],
  },
  {
    "ClearOrigin": "NULL",
  },
  {
    "BuyExecution": {
      "fees": {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 1,
        },
      },
      "weight_limit": {
        "Unlimited": "NULL",
      },
    },
  },
  {
    "DepositAsset": {
      "assets": {
        "Wild": {
          "AllCounted": 2,
        },
      },
      "beneficiary": {
        "interior": {
          "X1": [
            {
              "AccountKey20": {
                "key": "0x98891e5FD24Ef33A488A47101F65D212Ff6E650E",
                "network": null,
              },
            },
          ],
        },
        "parents": 0,
      },
      "max_assets": 0,
    },
  },
  {
    "SetTopic": "0xabcdef1234567890fedcba0987654321abcdef1234567890fedcba0987654321",
  },
]
`;

exports[`xcmPaymentApi.utils > getInstructions > should return ecosystem bridge instructions when isEcosystemBridge is true 1`] = `
[
  {
    "UniversalOrigin": {
      "GlobalConsensus": {
        "ByGenesis": undefined,
      },
    },
  },
  {
    "DescendOrigin": {
      "X1": [
        {
          "Parachain": 1001,
        },
      ],
    },
  },
  {
    "WithdrawAsset": [
      {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 1,
        },
      },
      {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 2,
        },
      },
    ],
  },
  {
    "ClearOrigin": "NULL",
  },
  {
    "BuyExecution": {
      "fees": {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 1,
        },
      },
      "weight_limit": {
        "Unlimited": "NULL",
      },
    },
  },
  {
    "DepositAsset": {
      "assets": {
        "Wild": {
          "AllCounted": 2,
        },
      },
      "beneficiary": {
        "interior": {
          "X1": [
            {
              "AccountKey20": {
                "key": "0x98891e5FD24Ef33A488A47101F65D212Ff6E650E",
                "network": null,
              },
            },
          ],
        },
        "parents": 0,
      },
      "max_assets": 0,
    },
  },
  {
    "SetTopic": "0xabcdef1234567890fedcba0987654321abcdef1234567890fedcba0987654321",
  },
]
`;

exports[`xcmPaymentApi.utils > getInstructions > should return reserve asset deposited instruction when isAssetReserveChain is false 1`] = `
[
  {
    "ReserveAssetDeposited": [
      {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 1,
        },
      },
      {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 2,
        },
      },
    ],
  },
  {
    "ClearOrigin": "NULL",
  },
  {
    "BuyExecution": {
      "fees": {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 1,
        },
      },
      "weight_limit": {
        "Unlimited": "NULL",
      },
    },
  },
  {
    "DepositAsset": {
      "assets": {
        "Wild": {
          "AllCounted": 2,
        },
      },
      "beneficiary": {
        "interior": {
          "X1": [
            {
              "AccountKey20": {
                "key": "0x98891e5FD24Ef33A488A47101F65D212Ff6E650E",
                "network": null,
              },
            },
          ],
        },
        "parents": 0,
      },
      "max_assets": 0,
    },
  },
  {
    "SetTopic": "0xabcdef1234567890fedcba0987654321abcdef1234567890fedcba0987654321",
  },
]
`;

exports[`xcmPaymentApi.utils > getInstructions > should return standard instructions when isEcosystemBridge is false 1`] = `
[
  {
    "WithdrawAsset": [
      {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 1,
        },
      },
      {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 2,
        },
      },
    ],
  },
  {
    "ClearOrigin": "NULL",
  },
  {
    "BuyExecution": {
      "fees": {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 1,
        },
      },
      "weight_limit": {
        "Unlimited": "NULL",
      },
    },
  },
  {
    "DepositAsset": {
      "assets": {
        "Wild": {
          "AllCounted": 2,
        },
      },
      "beneficiary": {
        "interior": {
          "X1": [
            {
              "AccountKey20": {
                "key": "0x98891e5FD24Ef33A488A47101F65D212Ff6E650E",
                "network": null,
              },
            },
          ],
        },
        "parents": 0,
      },
      "max_assets": 0,
    },
  },
  {
    "SetTopic": "0xabcdef1234567890fedcba0987654321abcdef1234567890fedcba0987654321",
  },
]
`;

exports[`xcmPaymentApi.utils > getInstructions > should return withdraw asset instruction when isAssetReserveChain is true 1`] = `
[
  {
    "WithdrawAsset": [
      {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 1,
        },
      },
      {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 2,
        },
      },
    ],
  },
  {
    "ClearOrigin": "NULL",
  },
  {
    "BuyExecution": {
      "fees": {
        "fun": {
          "Fungible": 1000000,
        },
        "id": {
          "id": 1,
        },
      },
      "weight_limit": {
        "Unlimited": "NULL",
      },
    },
  },
  {
    "DepositAsset": {
      "assets": {
        "Wild": {
          "AllCounted": 2,
        },
      },
      "beneficiary": {
        "interior": {
          "X1": [
            {
              "AccountKey20": {
                "key": "0x98891e5FD24Ef33A488A47101F65D212Ff6E650E",
                "network": null,
              },
            },
          ],
        },
        "parents": 0,
      },
      "max_assets": 0,
    },
  },
  {
    "SetTopic": "0xabcdef1234567890fedcba0987654321abcdef1234567890fedcba0987654321",
  },
]
`;

exports[`xcmPaymentApi.utils > getVersionedAssets > should put fee asset first when shouldTransferAssetPrecedeFeeAsset is false 1`] = `
[
  {
    "interior": "Here",
    "parents": 1,
  },
  {
    "interior": {
      "X1": [
        {
          "PalletInstance": 3,
        },
      ],
    },
    "parents": 1,
  },
]
`;

exports[`xcmPaymentApi.utils > getVersionedAssets > should put transfer asset first when shouldTransferAssetPrecedeFeeAsset is true 1`] = `
[
  {
    "interior": {
      "X1": [
        {
          "PalletInstance": 3,
        },
      ],
    },
    "parents": 1,
  },
  {
    "interior": "Here",
    "parents": 1,
  },
]
`;

exports[`xcmPaymentApi.utils > getVersionedAssets > should return only fee asset when assets are the same 1`] = `
[
  {
    "interior": "Here",
    "parents": 1,
  },
]
`;
